blockstore_caboose.go

```diff
@@ -0,0 +1,83 @@
+package main
+
+import (
+	"crypto/tls"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/filecoin-saturn/caboose"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+)
+
+func newCabooseBlockStore(orchestrator, loggingEndpoint string) (blockstore.Blockstore, error) {
+	var (
+		orchURL *url.URL
+		loggURL *url.URL
+		err     error
+	)
+
+	if orchestrator != "" {
+		orchURL, err = url.Parse(orchestrator)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if loggingEndpoint != "" {
+		loggURL, err = url.Parse(loggingEndpoint)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	saturnServiceClient := &http.Client{
+		Timeout: caboose.DefaultSaturnRequestTimeout,
+		Transport: &withUserAgent{
+			RoundTripper: &http.Transport{
+				DialContext: dialWithCachedDNS,
+			},
+		},
+	}
+
+	saturnRetrievalClient := &http.Client{
+		Timeout: caboose.DefaultSaturnRequestTimeout,
+		Transport: &withUserAgent{
+			RoundTripper: &http.Transport{
+				// Increasing concurrency defaults from http.DefaultTransport
+				MaxIdleConns:        1000,
+				MaxConnsPerHost:     100,
+				MaxIdleConnsPerHost: 100,
+				IdleConnTimeout:     90 * time.Second,
+
+				DialContext: dialWithCachedDNS,
+
+				// Saturn Weltschmerz
+				TLSClientConfig: &tls.Config{
+					// Saturn use TLS in controversial ways, which sooner or
+					// later will force them to switch away to different domain
+					// name and certs, in which case they will break us. Since
+					// we are fetching raw blocks and dont really care about
+					// TLS cert being legitimate, let's disable verification
+					// to save CPU and to avoid catastrophic failure when
+					// Saturn L1s suddenly switch to certs with different DNS name.
+					InsecureSkipVerify: true,
+					// ServerName:         "strn.pl",
+				},
```

Comment:
@guanzo is it possible to disable redirect from `http://` to `https://` if request is for CAR or a Block? These request types do not need TLS protection.

We would like to run tests and see if paying the cost of TLS gives us any perf benefit at all.

Reply:
some transport encryption so that a passive network observer can't see what CIDs are being accessed seems still useful

Reply:
I dont see how, everything is in the clear text on DHT, and here you get CIDs comigled together across all gateway users. 

Until we have fully privacy-preserving DHT, TLS here is just security theatre.

Reply:
* these queries are largely going to hit the L1 cache where they won't go back to the DHT - this lets you hide which CIDs are popular at the gateway
* arguing 'we can't improve privacy anywhere until we have it everywhere' doesn't seem like a winning strategy either. Unless this is a _huge_ reduction in performance it doesn't seem like this hurts.

Reply:
Ack. Even if we don't want to do this, measuring how much it _could buy us_ feels like a useful thing to know.
We can't measure it without disabling redirect I mentioned.  But ok to park this for now.

Reply:
> is it possible to update L1s and disable redirect from http:// to https:// if request is for verifiable CAR or a Block?
doable, ya. Although if the perf difference is significant, maybe this is something you can benchmark locally with a saturn fork?

Reply:
I won't be able to replicate real world traffic locally, but we have such traffic mirrored  on bifrost-gw-staging env.

Reply:
> some transport encryption so that a passive network observer can't see what CIDs are being accessed seems still useful

The code I sent disable HTTP not TLS, you still do TLS above HTTP1.1.

Reply:
@lidel should be disabled in prod now https://github.com/filecoin-saturn/L1-node/commit/108e0c6d3fd7e6dc1a1b9b9426a76e5900c0883c